### PR TITLE
Unit tests for reply-by-email feature functions in comment model

### LINKED ISSUE #3435: Add remaining unit tests for reply-by-email feature functions in comment model
**We are preparing to participate in Google Code-in, and have reserved this issue for participants in GCI - but we'd love to have your help with another one! Please check out https://code.publiclab.org to see more.**

There are some functions in `reply-by-email` feature in comment model for which there are no unit tests.
We can add them in test/unit/comment_test.rb

https://github.com/publiclab/plots2/blob/faa03523b35ccaddd26f4bfc00a2c0bcce591b5b/app/models/comment.rb#L224

https://github.com/publiclab/plots2/blob/faa03523b35ccaddd26f4bfc00a2c0bcce591b5b/app/models/comment.rb#L261

https://github.com/publiclab/plots2/blob/faa03523b35ccaddd26f4bfc00a2c0bcce591b5b/app/models/comment.rb#L291

https://github.com/publiclab/plots2/blob/faa03523b35ccaddd26f4bfc00a2c0bcce591b5b/app/models/comment.rb#L299

https://github.com/publiclab/plots2/blob/faa03523b35ccaddd26f4bfc00a2c0bcce591b5b/app/models/comment.rb#L315

### PR DESCRIPTION
**Pull request by dinaelhanan**
Fixes #3435 
feel free to add to this, and sorry for the comments spam, I'm trying to document my learning process through this in case any other people who are new to this want to check it out :)

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] ask `@publiclab/reviewers` for help, in a comment below

### CODE REVIEW

**Comment by oorjitchowdhary** [comment]
@dinaelhanan Thanks for your PR.
I believe we have to add the unit tests for all the 5 functions mentioned in #3435 in order to complete the issue..

**Comment by dinaelhanan** [comment]
@oorjitchowdhary It's a WIP, I'll add that to the PR name, I'm looking for help on writing unit tests, as this is my first one

**Comment by oorjitchowdhary** [comment]
Oh OK sure..

**Comment by dinaelhanan** [comment]
@oorjitchowdhary I would love your help though! See as I was getting errors, I followed [this](https://publiclab.org/wiki/contributing-to-public-lab-software#Rewinding+the+master+branch) to try and fix it. Let me know if I did something wrong

**Comment by dinaelhanan** [comment]
need help! @publiclab/reviewers

**Comment by dinaelhanan** [comment]
I think I figured out the issue, travis build says:
`ERROR["test_should_have_gmail_quote", #<Minitest::Reporters::Suite:0x000055d9a0e17c40 @name="CommentTest">, 27.03831857199998]
 test_should_have_gmail_quote#CommentTest (27.04s)
NoMethodError:         NoMethodError: undefined method `gmail_quote_present?' for #<CommentTest:0x000055d99f93a390>
            test/unit/comment_test.rb:20:in `block in <class:CommentTest>'`

**Comment by JonathanXu1** [comment]
I've never written unit tests before as well. If you could let me know if you figure it out that would be awesome. Good luck!

**Comment by dinaelhanan** [comment]
@JonathanXu1 Definitely! It's a learning process for me that I'm hoping to document well so that other people can use it too!

**Comment by dinaelhanan** [comment]
Update: I realized that I was getting the NoMethodError for using the method name "self.gmail_quote_present?" without "comment." before it... facepalm!

**Comment by gauravano** [comment]
You can run a particular test too - https://stackoverflow.com/questions/1506780/how-to-run-a-single-test-from-a-rails-test-suite

**Comment by dinaelhanan** [comment]
@gauravano that looks helpful! I'll do that and report back here

**Comment by gauravano** [comment]
`gmail_quote_present` function is inaccessible that's why test is failing.

**Comment by gauravano** [comment]
I was going through the function in question and I saw that function is also expecting an argument. See app/models/comment.rb

**Comment by dinaelhanan** [comment]
@gauravano it is? How would I do that in the test?

**Comment by gauravano** [comment]
I am done for the day. But, before going, here are some tips:
1. Whichever code editor you are using, find occurrences of
gmail_quote_present through it. I use Rubymine.
2. After that try to generate the argument in test or if you are just
testing a line for now, may be target test function for which you are
testing might already have this argument present?

hope you'll get it to work. If not, you'll get help tomorrow from jeff or
namangupta , people are enjoying Sunday :)

**Comment by dinaelhanan** [comment]
@gauravano ok thanks, I'll try that and keep updating! :+1:

**Comment by dinaelhanan** [comment]
@namangupta01 hey! I'm working on this and I'm having trouble, any chance you can help out? Thanks!

**Comment by namangupta01** [comment]
Hey, where are you facing problem?

**Comment by namangupta01** [comment]
Here in this function https://github.com/publiclab/plots2/blob/ff870dea11781ab5ed50a806a55f362210f3a638/app/models/comment.rb#L291
the argument is an object of HTML class of Nokogiri gem. So you have to pass the same object there.
For getting this, you have to use the `fixture` and 
 first make the object of `Mail` class as shown here https://github.com/publiclab/plots2/blob/ff870dea11781ab5ed50a806a55f362210f3a638/test/integration/incoming_mail_parsing_test/gmail_parsing_test.rb#L5 and then convert this mail object to `Nokogiri::HTML` object as done in this line https://github.com/publiclab/plots2/blob/ff870dea11781ab5ed50a806a55f362210f3a638/app/models/comment.rb#L264
Now pass this argument and you are good to go. And yes, don't forget to `require` the related gem in the test file. :)

**Comment by dinaelhanan** [comment]
@gauravano I should have seen that! I have done most of the above... But how do I require the related gem in the test file?

**Comment by ValentinaTironi** [comment]
maybe `require 'nokogiri'`

**Comment by namangupta01** [comment]
Yup @ValentinaTironi is right.
https://github.com/publiclab/plots2/blob/ff870dea11781ab5ed50a806a55f362210f3a638/test/integration/incoming_mail_parsing_test/gmail_parsing_test.rb#L4

**Comment by dinaelhanan** [comment]
@ValentinaTironi @namangupta01  that sounds great! Thanks so much for your guidance, this was my first time writing the unit test! I will do the same for the remaining ones in #3435!

File: test/unit/comment_test.rb [hunk 1]
```diff
@@ -16,7 +16,12 @@ class CommentTest < ActiveSupport::TestCase
       assert comments.length > 0
     end
   end
-
+  test 'should have gmail quote' do
+    mail = Mail.read('test/fixtures/incoming_test_emails/gmail/incoming_gmail_email.eml') 
+    mail_doc = Nokogiri::HTML(mail.html_part.body.decoded) # To parse the mail to extract comment content and reply content 
+    gmail_quote = Comment.gmail_quote_present?(".gmail_quote")
```
Comment (namangupta01) [inline test/unit/comment_test.rb]: @dinaelhanan You forgot to change the argument :joy:

Comment (dinaelhanan) [reply test/unit/comment_test.rb]: ooh woops!

**Comment by namangupta01** [comment]
I think the test should pass this time.

**Comment by dinaelhanan** [comment]
Hopefully! @namangupta01 please check out my task on google code-in when it's finished if you can!

**Comment by namangupta01** [comment]
@dinaelhanan You added the nokogiri gem in the wrong file. Add it in the file where you are using it.

File: test/unit/comment_test.rb [hunk 2]
```diff
@@ -16,7 +16,12 @@ class CommentTest < ActiveSupport::TestCase
       assert comments.length > 0
     end
   end
-
+  test 'should have gmail quote' do
```
Comment (namangupta01) [inline test/unit/comment_test.rb]: Add after this line : 
require 'mail'
require 'nokogiri'

Comment (dinaelhanan) [reply test/unit/comment_test.rb]: Oh I didn't know that... Thanks for pointing that out!

**Comment by dinaelhanan** [comment]
@namangupta01 yay! All green. Can you check out my 
[task](https://codein.withgoogle.com/dashboard/task-instances/4593095225114624/)

**Comment by jywarren** [comment]
Very nice!!! Great teamwork!!!

**Pull request merged**